DataIndex addition on SourceTable coalesce breaks if index columns are dropped via view

```
heduler-Concurrent-1 | i.d.s.s.SessionState      | Internal Error '41587ce6-da87-4c6f-8d75-33d2d33ca1fc' io.deephaven.engine.table.impl.locations.TableDataException: Error initializing location sizes
	at io.deephaven.engine.table.impl.SourceTable.lambda$initializeLocationSizes$3(SourceTable.java:197)
	at io.deephaven.engine.table.impl.perf.QueryPerformanceRecorder.withNugget(QueryPerformanceRecorder.java:353)
	at io.deephaven.engine.table.impl.SourceTable.initializeLocationSizes(SourceTable.java:191)
	at io.deephaven.engine.table.impl.SourceTable.initialize(SourceTable.java:117)
	at io.deephaven.engine.table.impl.SourceTable.doCoalesce(SourceTable.java:267)
	at io.deephaven.engine.table.impl.SourceTable.doCoalesce(SourceTable.java:33)
	at io.deephaven.engine.table.impl.UncoalescedTable.coalesce(UncoalescedTable.java:78)
	at io.deephaven.engine.table.impl.UncoalescedTable.flatten(UncoalescedTable.java:447)
	at io.deephaven.server.table.ops.FlattenTableGrpcImpl.create(FlattenTableGrpcImpl.java:29)
	at io.deephaven.server.table.ops.FlattenTableGrpcImpl.create(FlattenTableGrpcImpl.java:16)
	at io.deephaven.server.table.ops.TableServiceGrpcImpl$BatchExportBuilder.doExport(TableServiceGrpcImpl.java:757)
	at io.deephaven.server.session.SessionState$ExportObject.doExport(SessionState.java:995)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at io.deephaven.server.runner.scheduler.SchedulerModule$ThreadFactory.lambda$newThread$0(SchedulerModule.java:100)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:209)
	at java.base/java.util.ImmutableCollections$MapN.get(ImmutableCollections.java:1232)
	at io.deephaven.engine.table.impl.indexer.DataIndexer$DataIndexCache.traverse(DataIndexer.java:443)
	at io.deephaven.engine.table.impl.indexer.DataIndexer$DataIndexCache.add(DataIndexer.java:507)
	at io.deephaven.engine.table.impl.indexer.DataIndexer.addDataIndex(DataIndexer.java:304)
	at io.deephaven.engine.table.impl.sources.regioned.RegionedColumnSourceManager.initialize(RegionedColumnSourceManager.java:251)
	at io.deephaven.engine.table.impl.SourceTable.lambda$initializeLocationSizes$3(SourceTable.java:195)
	... 18 more
```

We need to refuse to create indexes if we cannot find the appropriate column sources.

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/dataindex/TableBackedDataIndex.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/dataindex/TableBackedDataIndex.java
@@ -5,6 +5,7 @@ package io.deephaven.engine.table.impl.dataindex;
 
 import io.deephaven.api.ColumnName;
 import io.deephaven.base.verify.Assert;
+import io.deephaven.base.verify.Require;
 import io.deephaven.engine.context.ExecutionContext;
 import io.deephaven.engine.liveness.LivenessScopeStack;
 import io.deephaven.engine.table.ColumnSource;
@@ -52,7 +53,7 @@ public class TableBackedDataIndex extends AbstractDataIndex {
     public TableBackedDataIndex(
             @NotNull final QueryTable sourceTable,
             @NotNull final String... keyColumnNames) {
-        this.keyColumnNames = List.of(keyColumnNames);
+        this.keyColumnNames = List.of(Require.elementsNeqNull(keyColumnNames, "keyColumnNames"));
 
         // Create an in-order reverse lookup map for the key column names.
         keyColumnNamesByIndexedColumn = Collections.unmodifiableMap(

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/MergedDataIndex.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/MergedDataIndex.java
@@ -75,8 +75,9 @@ class MergedDataIndex extends AbstractDataIndex {
             @NotNull final String[] keyColumnNames,
             @NotNull final ColumnSource<?>[] keySources,
             @NotNull final RegionedColumnSourceManager columnSourceManager) {
-
         Require.eq(keyColumnNames.length, "keyColumnNames.length", keySources.length, "keySources.length");
+        Require.elementsNeqNull(keyColumnNames, "keyColumnNames");
+        Require.elementsNeqNull(keySources, "keySources");
 
         this.keyColumnNames = List.of(keyColumnNames);
         this.columnSourceManager = columnSourceManager;

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/PartitioningColumnDataIndex.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/PartitioningColumnDataIndex.java
@@ -5,6 +5,7 @@ package io.deephaven.engine.table.impl.sources.regioned;
 
 import gnu.trove.map.hash.TObjectIntHashMap;
 import io.deephaven.base.verify.Assert;
+import io.deephaven.base.verify.Require;
 import io.deephaven.engine.rowset.RowSequence;
 import io.deephaven.engine.rowset.RowSet;
 import io.deephaven.engine.rowset.RowSetBuilderRandom;
@@ -73,7 +74,9 @@ class PartitioningColumnDataIndex<KEY_TYPE> extends AbstractDataIndex {
             @NotNull final String keyColumnName,
             @NotNull final ColumnSource<KEY_TYPE> keySource,
             @NotNull final RegionedColumnSourceManager columnSourceManager) {
-        this.keyColumnName = keyColumnName;
+        this.keyColumnName = Require.neqNull(keyColumnName, "keyColumnName");
+        Require.neqNull(keySource, "keySource");
+        Require.neqNull(columnSourceManager, "columnSourceManager");
 
         keyColumnNamesByIndexedColumn = Map.of(keySource, keyColumnName);
 

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/RegionedColumnSourceManager.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/RegionedColumnSourceManager.java
@@ -245,6 +245,12 @@ public class RegionedColumnSourceManager extends LivenessArtifact implements Col
                 final ColumnSource<?>[] keySources = Arrays.stream(keyColumnNames)
                         .map(columnSources::get)
                         .toArray(ColumnSource[]::new);
+                if (Arrays.stream(keySources).anyMatch(Objects::isNull)) {
+                    // If any of the key sources are missing (e.g. because the user supplied a TableDefinition that
+                    // excludes them, or has used a RedefinableTable operation to drop them), we can't create the
+                    // DataIndex.
+                    continue;
+                }
                 final DataIndex mergedIndex = new MergedDataIndex(keyColumnNames, keySources, this);
                 retainedDataIndexes.add(mergedIndex);
                 // Not refreshing, so no need to manage mergedIndex

--- a/engine/table/src/test/java/io/deephaven/engine/table/impl/locations/impl/TestSourceTableDataIndexes.java
+++ b/engine/table/src/test/java/io/deephaven/engine/table/impl/locations/impl/TestSourceTableDataIndexes.java
@@ -174,6 +174,21 @@ public class TestSourceTableDataIndexes {
     }
 
     @Test
+    public void testDroppedIndexColumn() {
+        final Table raw = TableTools.emptyTable(26 * 10 * 1000).update("Part=String.format(`%04d`, (long)(ii/1000))",
+                "Sym=(char)('A' + ii % 26)", "Other=ii");
+        DataIndexer.getOrCreateDataIndex(raw, "Sym");
+
+        final String path =
+                dataDirectory.getAbsolutePath() + File.separator + "TestTable2" + File.separator + PARQUET_FILE_NAME;
+
+        ParquetTools.writeTable(raw, path);
+
+        TestCase.assertFalse(DataIndexer.hasDataIndex(
+                ParquetTools.readTable(path).dropColumns("Sym").coalesce(), "Sym"));
+    }
+
+    @Test
     public void testParallelCollection() {
         final List<Integer> observedOrder = Collections.synchronizedList(new ArrayList<>());
         final int[] intArray = IntStream.range(0, 10000).parallel().peek(observedOrder::add).toArray();

--- a/engine/table/src/test/java/io/deephaven/engine/table/impl/locations/impl/TestSourceTableDataIndexes.java
+++ b/engine/table/src/test/java/io/deephaven/engine/table/impl/locations/impl/TestSourceTableDataIndexes.java
@@ -102,7 +102,6 @@ public class TestSourceTableDataIndexes {
 
         final TableDefinition partitionedMissingDataDefinition = TableDefinition.of(
                 ColumnDefinition.ofString("Part").withPartitioning(),
-                ColumnDefinition.ofChar("Sym"),
                 ColumnDefinition.ofLong("Other"));
 
         final String tableName = "TestTable";

